pkg/alertmanager/multitenant.go

```diff
@@ -838,6 +848,57 @@ func (am *MultitenantAlertmanager) GetStatusHandler() StatusHandler {
 	}
 }
 
+// deleteObsoleteLocalFiles finds local files that we no longer need, and deletes them.
+func (am *MultitenantAlertmanager) deleteObsoleteLocalFiles() {
+	userFilesToDelete := am.getSnapshotFilesPerUser()
+
+	am.alertmanagersMtx.Lock()
+	for userID := range am.alertmanagers {
+		delete(userFilesToDelete, userID) // Don't delete files for users we're managing.
+	}
+	am.alertmanagersMtx.Unlock()
+
+	// And delete remaining files.
+	for userID, files := range userFilesToDelete {
+		for _, fn := range files {
+			if err := os.Remove(fn); err != nil {
```

Comment:
There's no guarantee that the alertmanager has not been created in the meanwhile (due to an HTTP request).

Reply:
That is true, but if AM is running again, it will also write the file to disk. It should not be a problem, from what I understand. (Except there is a time between delete and recreation of file, when possible crash will lose AM state for given user)

Reply:
I also share Marco's concerns but also understand that the code, as written will just log failures and move on.

Reply:
There is another way to avoid this race, by adding random suffix to the state file. When AM is created, and it finds nflog:<user>.<suffix> (or silences), it will rename it to use random suffix before using. If rename succeeds, we have just saved the file from deletion – cleanup will try to delete old name and fail. If rename fails, then file was already gone, and new one is created. (Note that rename is atomic operation on posix filesystems)

Reply:
I think this may be a good idea. I'm still wondering if a per-alertmanager lock+ state may be better (eg. do not allow to create a new alertmanager for a user if another alertmanager for that user is still stopping). It would help up having a more sane locking approach and would implicitly solve the race condition we're discussing here too.

> When AM is created, and it finds nflog:<user>.<suffix> (or silences), it will rename it to use new random suffix before using.

I don't even think we need to rename. We can generate a new random suffix when we create a new alertmanager. Collisions are so rare that we can really ignore it.

Reply:
> I don't even think we need to rename. We can generate a new random suffix when we create a new alertmanager. Collisions are so rare that we can really ignore it.

The point of the rename was to avoid deleting previous state file, if cleanup hasn't finished yet -- this would preserve it. But I agree that in practice this will likely not happen very often (race between stopping and starting new alertmanager for tenant).

Reply:
> The point of the rename was to avoid deleting previous state file, if cleanup hasn't finished yet -- this would preserve it. But I agree that in practice this will likely not happen very often (race between stopping and starting new alertmanager for tenant).

Sure. I get it and I think we should fix it. The random suffix should work for this use case. Having a more sane approach to alertmanager locking (and its state) could a definitive solution. That was my comment.